REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [96]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [97]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [6]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

In [28]:
bus['business_id'].nunique()

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [36]:
bus['name'].value_counts()[:5]

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [113]:
bus['postal_code']

0       94104
1       94109
2       94104
3       94133
4       94110
        ...  
6310    94103
6311    94127
6312    94116
6313    94112
6314    94110
Name: postal_code, Length: 6315, dtype: object

In [ ]:
# they are stored as objects as it says above

#### Question 3b

What are the unique values of postal_code?

In [41]:
bus['postal_code'].value_counts()

94110        585
94103        558
94102        458
94107        451
94133        417
94109        383
94111        282
94122        270
94118        243
94115        238
94114        230
94105        226
94108        224
94117        197
94112        190
94123        169
94124        164
94121        161
94132        134
94104        133
94116         93
94134         74
94127         71
94131         48
94158         17
94130          8
94143          5
CA             3
Ca             3
94101          2
94013          2
94129          2
94014          2
941033148      1
94602          1
94621          1
94188          1
00000          1
94609          1
95105          1
941            1
94120          1
941102019      1
94545          1
94066          1
92672          1
Name: postal_code, dtype: int64

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [111]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]
bus_valid_index = bus['postal_code'].isin(validZip)
bus_valid = bus[bus_valid_index]
print(bus_valid.reset_index())

      index  business_id                                name  \
0         0           10                    TIRAMISU KITCHEN   
1         1           19               NRGIZE LIFESTYLE CAFE   
2         2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3         3           31      NORMAN'S ICE CREAM AND FREEZES   
4         4           45                 CHARLIE'S DELI CAFE   
...     ...          ...                                 ...   
5994   6310        90234                ACCESS SFUSD THE ARC   
5995   6311        90236                      PHO DONG HUONG   
5996   6312        90237                    WIN'S RESTAURANT   
5997   6313        90238                    USA POWER MARKET   
5998   6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIF

In [114]:
# this is with the code that was provided, I just called it 'second try' because I wasn't sure if you wanted us to print the list
# with only the postal codes that are valid, or print all the businesses but leave the invalid zip codes blank

bus_valid_secondtry = bus[bus['postal_code'].isin(validZip)]
print(bus_valid_secondtry)

      business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENW

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [115]:
bus_valid[pd.isnull(bus_valid['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [209]:
newdf = pd.DataFrame()
newdf['postal_code'] = validZip
newdf.index = validZip
#newdf.loc[:, 'postal_code'] = list(bus_valid['postal_code'])
null = bus_valid[pd.isnull(bus_valid['longitude'])]
newdf.loc[:, 'null_lon'] = null.groupby('postal_code')['business_id'].nunique()

notnull = bus_valid[pd.notnull(bus_valid['longitude'])]
newdf.loc[:, 'not_null_lon'] = notnull.groupby('postal_code')['business_id'].nunique()

print(newdf)

      postal_code  null_lon  not_null_lon
94102       94102       196           262
94103       94103       252           306
94104       94104        59            74
94105       94105       100           126
94107       94107       247           204
94108       94108        82           142
94109       94109       146           237
94110       94110       251           334
94111       94111       115           167
94112       94112        59           131
94114       94114       108           122
94115       94115        76           162
94116       94116        35            58
94117       94117        67           130
94118       94118       100           143
94121       94121        55           106
94122       94122       123           147
94123       94123        53           116
94124       94124        88            76
94127       94127        28            43
94131       94131        14            34
94132       94132        65            69
94133       94133       135       

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [214]:
newdf['prop_miss_pc'] = newdf['null_lon'] / newdf['not_null_lon']
print(newdf)

top = newdf.nsmallest(n=5, columns=['prop_miss_pc'])
top.head(5)

      postal_code  null_lon  not_null_lon  prop_miss_pc
94102       94102       196           262      0.748092
94103       94103       252           306      0.823529
94104       94104        59            74      0.797297
94105       94105       100           126      0.793651
94107       94107       247           204      1.210784
94108       94108        82           142      0.577465
94109       94109       146           237      0.616034
94110       94110       251           334      0.751497
94111       94111       115           167      0.688623
94112       94112        59           131      0.450382
94114       94114       108           122      0.885246
94115       94115        76           162      0.469136
94116       94116        35            58      0.603448
94117       94117        67           130      0.515385
94118       94118       100           143      0.699301
94121       94121        55           106      0.518868
94122       94122       123           147      0

postal_code  null_lon  not_null_lon  prop_miss_pc
94131       94131        14            34      0.411765
94112       94112        59           131      0.450382
94123       94123        53           116      0.456897
94115       94115        76           162      0.469136
94133       94133       135           282      0.478723

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [101]:
len(ins)

15430

In [102]:
ins['business_id'].nunique()

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [156]:
ins['type'].value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

In [198]:
#ins2['type'] = pd.ins.groupby('business_id')['type'].apply(lambda x > 10)


#ins2 = ins[ins['type'] > 10].groupby(['business_id']).nunique()
#ins['type']

TypeError: '>' not supported between instances of 'str' and 'int'

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [224]:
ins.date = ins.date.astype(str)
ins.loc[:,'year'] = ins['date'].str[:4]
ins

#hi Irene, i think my data is messed up, sorry about that... this code works, but my data is changed for some reason

business_id  score                           date     type  year  \
0               10     82  1970-01-01 00:00:00.020160503  routine  1970   
1               10     94  1970-01-01 00:00:00.020140729  routine  1970   
2               10     92  1970-01-01 00:00:00.020140114  routine  1970   
3               19     94  1970-01-01 00:00:00.020160513  routine  1970   
4               19     94  1970-01-01 00:00:00.020141110  routine  1970   
...            ...    ...                            ...      ...   ...   
15425        89515     98  1970-01-01 00:00:00.020161103  routine  1970   
15426        89701    100  1970-01-01 00:00:00.020161206  routine  1970   
15427        89790     96  1970-01-01 00:00:00.020161129  routine  1970   
15428        89803    100  1970-01-01 00:00:00.020161215  routine  1970   
15429        89900     93  1970-01-01 00:00:00.020161206  routine  1970   

         month  day day_of_week  
0      January    1    Thursday  
1      January    1    Thursday  
2      January    1    Thursday  
3      January    1    Thursday  
4      January    1    Thursday  
...        ...  ...         ...  
15425  January    1    Thursday  
15426  January    1    Thursday  
15427  January    1    Thursday  
15428  January    1    Thursday  
15429  January    1    Thursday  

[15430 rows x 8 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [225]:
ins['year'].value_counts()

1970    15430
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [138]:
bus_ins = pd.merge(bus, ins)
bus_ins.head()

business_id                   name                       address  \
0           10       TIRAMISU KITCHEN                 033 BELDEN PL   
1           10       TIRAMISU KITCHEN                 033 BELDEN PL   
2           10       TIRAMISU KITCHEN                 033 BELDEN PL   
3           19  NRGIZE LIFESTYLE CAFE  1200 VAN NESS AVE, 3RD FLOOR   
4           19  NRGIZE LIFESTYLE CAFE  1200 VAN NESS AVE, 3RD FLOOR   

            city state postal_code   latitude   longitude  phone_number  \
0  San Francisco    CA       94104  37.791116 -122.403816  +14154217044   
1  San Francisco    CA       94104  37.791116 -122.403816  +14154217044   
2  San Francisco    CA       94104  37.791116 -122.403816  +14154217044   
3  San Francisco    CA       94109  37.786848 -122.421547  +14157763262   
4  San Francisco    CA       94109  37.786848 -122.421547  +14157763262   

   score      date     type  
0     82  20160503  routine  
1     94  20140729  routine  
2     92  20140114  routine  
3     94  20160513  routine  
4     94  20141110  routine

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [155]:
lowest = bus_ins.loc[:,['name', 'address', 'score']].nsmallest(n=20, columns=['score'])
lowest.head(20)

name                     address  \
10630                        ABC BAKERY CAFE             650 JACKSON ST    
6356                         IMPERIAL PALACE          818 WASHINGTON ST    
9237          YUMMY DIM SUM & FAST FOOD, LLC            930 STOCKTON ST    
11657                GOLDEN RIVER RESTAURANT            5827 GEARY BLVD    
2431                        YEE'S RESTAURANT             1131 GRANT AVE    
2451                             LES JOULINS                44 ELLIS ST    
2720             BASIL THAI RESTAURANT & BAR             1175 FOLSOM ST    
5907                           BODEGA BISTRO              607 LARKIN ST    
6703                     BRISTOL FARMS #2317           845 MARKET ST 010   
8612        OZONE THAI RESTAURANT AND LOUNGE                598 02ND ST    
12016                        MISS SAIGON LLC                100 06TH ST    
3368                         GOURMET KITCHEN           1051 STOCKTON ST    
7739                              FIRST CAKE            5901 GEARY BLVD    
11600                          SA BEANG THAI          312 DIVISADERO ST    
11614                             UNCLE CAFE              65 WAVERLY PL    
14250  LE SOLEIL INTERNATIONAL HOLDINGS, INC          133 CLEMENT  ST.     
14296                           CRAZY PEPPER          2257 SAN JOSE AVE    
14411                HARBOR VILLA RESTAURANT  5238 DIAMOND HEIGHTS BLVD    
15317                              POKI TIME            2101 LOMBARD ST    
861                          TAQUERIA LA PAZ                305 TURK ST    

       score  
10630     46  
6356      47  
9237      50  
11657     52  
2431      53  
2451      53  
2720      54  
5907      54  
6703      54  
8612      54  
12016     54  
3368      55  
7739      55  
11600     55  
11614     55  
14250     55  
14296     55  
14411     55  
15317     55  
861       56

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!